hal/sai_api.h

```diff
@@ -0,0 +1,153 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_SAI_API_H
+#define MBED_SAI_API_H
+
+#include "device.h"
+
+#if DEVICE_SAI
+
+#ifdef __cplusplus
+extern "C" {
+#endif // __cplusplus
+
+/**
+ * \defgroup hal_sai Serial audio interface hal functions
+ * Low level interface to the serial audio interface of a target.
+ *
+ * # Defined behavior
+ * * Supports a subset of the possible configuration space - verified by ::sai_
+ * * Reports a failure (returns false) upon any invocation to an unsupported feature/parameter - verified by ::sai_
+ *
```

Comment:
I suspect that defined behavior is more detailed and provides requirements for each SAI HAL API:

> Reports a failure (returns false) upon any invocation to an unsupported feature/parameter.

Suggest to change to: 
- `sai_init()` returns SAI_RESULT_CONFIG_UNSUPPORTED/SAI_RESULT_CONFIG_MISMATCH`) if given configuration is unsupported/invalid. 

Additionally I would need to have strictly defined when the configuration is unsupported and when invalid, so I can write a test. On the other hand I'm not sure if we need `SAI_RESULT_CONFIG_MISMATCH`? Maybe in case when driver can not handle the given configuration should always return `SAI_RESULT_CONFIG_UNSUPPORTED`. That would be more predictable.

I would also add following additional requirements:
- `sai_init()` returns `SAI_RESULT_INVALID_PARAM` if at least one of the given parameters is undefined (NULL).
- `sai_init()` returns `SAI_RESULT_INIT_ERROR` if SAI has been already initialized.
- `sai_xfer()` transmits sample and returns `true` on success if SAI operates as transmitter.
- `sai_xfer()` receives sample and returns `true` on success if SAI operates as  receiver.
- `sai_xfer()` can simultaneously transmit and receive sample if SAI operates as transmitter and receiver.
- `sai_xfer()` returns false if at least one of given parameters is undefined (NULL).
- `sai_xfer()` returns false if sample has not been successfully transmitted and SAI operates as transmitter.
- `sai_xfer()` returns false if no sample has been received and SAI operates as receiver.
- what `sai_xfer()` shell return in case when SAI operates as transmitter and receiver and sample has been successfully transmitted, but no sample has been received, or vice versa? Maybe we should consider to extend the status or split back `sai_xfer()` into `sai_write()` and `sai_read()` ?
- `sai_free()` releases & de-initializes the referenced peripherals.
- after `sai_free()`  SAI can be re-initialized by means of `sai_init()`.
- `sai_free()` does nothing if given parameter is undefined (NULL).

This is only a proposition based on which I would like to create the test. Could you please review this and provide your opinion?

Reply:
Hi,
- `SAI_RESULT_CONFIG_UNSUPPORTED` is for when the device can *NEVER* support this config
- `SAI_RESULT_CONFIG_MISMATCH` is for when the device is not able to support this config at this very moment because some other active device prevents it from doing so.

As a SAI cannot be transmitter **AND** receiver at the same time, I propose this instead :
- `sai_init()` returns `SAI_RESULT_INVALID_PARAM`  if at least one of the given parameters is undefined (NULL).
- `sai_init()` returns `SAI_RESULT_ALREADY_INITIALIZED` if SAI is already initialized
- `sai_free()` does nothing if passed a NULL pointer.
- `sai_free()` de-initialized & un-clock unused part of the device.
- a device/block can be reinitialized via `sai_init()` after being `sai_free()`d.

if the device is a *receiver* :
- `sai_xfer()` returns false if the `sai_t` object is NULL
- `sai_xfer()` if `psample` is NULL : it pops 1 sample from the FiFo and returns true
- `sai_xfer()` if `psample` is not NULL : it pops 1 sample from the FiFo, stores it to the address pointed by `psample`,  and returns true
-`sai_xfer()` returns false if there's no sample in the FiFo.

if the device is a *transmitter* :
- `sai_xfer()` returns false if the `sai_t` object is NULL
-`sai_xfer()` returns false if the fifo is full and `*psample` could not be pushed.
- `sai_xfer()` if `psample` is NULL : it pushes one '0' sample to the FiFo and returns true
- `sai_xfer()` if `psample` is not NULL : it pushes the pointed sample to the FiFo and returns true